### PR TITLE
Include response data in 'request over websocket' failures.

### PR DESCRIPTION
**Pull request by charlesmchen-signal**
PTAL @michaelkirk

### CODE REVIEW

File: SignalServiceKit/src/Messages/OWSMessageSender.m [hunk 1]
```diff
@@ -1009,6 +1010,7 @@ - (void)sendMessageToService:(TSOutgoingMessage *)message
             failure:^(NSURLSessionDataTask *task, NSError *error) {
                 NSHTTPURLResponse *response = (NSHTTPURLResponse *)task.response;
                 NSInteger statusCode = response.statusCode;
+                NSData *_Nullable responseData = error.userInfo[AFNetworkingOperationFailingURLResponseDataErrorKey];
```
Comment (charlesmchen-signal) [inline SignalServiceKit/src/Messages/OWSMessageSender.m]: AFNetworking exposes error response bodies using `AFNetworkingOperationFailingURLResponseDataErrorKey`, but we need to include them in the failure callback.

File: SignalServiceKit/src/Messages/OWSMessageSender.m [hunk 2]
```diff
@@ -1130,32 +1131,39 @@ - (void)messageSendDidFail:(TSOutgoingMessage *)message
...
-                [NSJSONSerialization JSONObjectWithData:responseData options:0 error:&error];
-            if (error) {
+            NSError *_Nullable error = nil;
+            NSDictionary *_Nullable responseJson = nil;
+            if (responseData) {
+                responseJson = [NSJSONSerialization JSONObjectWithData:responseData options:0 error:&error];
+            }
+            if (error || !responseJson) {
                 OWSProdError([OWSAnalyticsEvents messageSenderErrorCouldNotParseMismatchedDevicesJson]);
                 [error setIsRetryable:YES];
```
Comment (charlesmchen-signal) [inline SignalServiceKit/src/Messages/OWSMessageSender.m]: I think it's pretty unlikely that retrying would succeed, but I won't change this.

File: SignalServiceKit/src/Network/WebSockets/TSSocketManager.m [hunk 3]
```diff
@@ -117,12 +117,14 @@ - (void)didFailWithStatusCode:(NSInteger)statusCode error:(NSError *)error
         self.hasCompleted = YES;
     }
 
+    DDLogError(@"%@ %s didFailWithStatusCode: %zd, %@", self.logTag, __PRETTY_FUNCTION__, statusCode, error);
```
Comment (charlesmchen-signal) [inline SignalServiceKit/src/Network/WebSockets/TSSocketManager.m]: We definitely want this logging.

File: SignalServiceKit/src/Messages/OWSMessageSender.m [hunk 4]
```diff
@@ -1442,13 +1454,12 @@ - (void)saveGroupMessage:(TSOutgoingMessage *)message inThread:(TSThread *)threa
...
 // Called when the server indicates that the devices no longer exist - e.g. when the remote recipient has reinstalled.
-- (void)handleStaleDevicesWithResponse:(NSData *)responseData
-                           recipientId:(NSString *)identifier
-                            completion:(void (^)(void))completionHandler
+- (void)handleStaleDevicesWithResponseJson:(NSDictionary *)responseJson
+                               recipientId:(NSString *)identifier
+                                completion:(void (^)(void))completionHandler
 {
     dispatch_async([OWSDispatch sendingQueue], ^{
-        NSDictionary *serialization = [NSJSONSerialization JSONObjectWithData:responseData options:0 error:nil];
```
Comment (michaelkirk-signal) [inline SignalServiceKit/src/Messages/OWSMessageSender.m]: glad to see this got moved up a level. 👍

File: SignalServiceKit/src/Network/WebSockets/TSSocketManager.h [hunk 5]
```diff
@@ -18,7 +18,7 @@ typedef NS_ENUM(NSUInteger, SocketManagerState) {
 
 typedef void (^TSSocketMessageSuccess)(id _Nullable responseObject);
 // statusCode is zero by default, if request never made or failed.
-typedef void (^TSSocketMessageFailure)(NSInteger statusCode, NSError *error);
+typedef void (^TSSocketMessageFailure)(NSInteger statusCode, NSData *_Nullable responseBody, NSError *error);
```
Comment (michaelkirk-signal) [inline SignalServiceKit/src/Network/WebSockets/TSSocketManager.h]: naming nit: maybe `responseData` instead of `responseBody`, but TIOLI

Comment (charlesmchen-signal) [reply SignalServiceKit/src/Network/WebSockets/TSSocketManager.h]: Done.

**Pull request merged**